lib/rouge/lexers/xpath.rb

```diff
@@ -0,0 +1,298 @@
+
+      # Lexical states:
+      # https://www.w3.org/TR/xquery-xpath-parsing/#XPath-lexical-states
+      # https://lists.w3.org/Archives/Public/public-qt-comments/2004Aug/0127.html
+      # https://www.w3.org/TR/xpath-30/#id-revision-log
+      # https://www.w3.org/TR/xpath-31/#id-revision-log
+
+      state :root do
+        # Comments
+        rule commentStart, Comment, :comment
+
+        # Literals
+        rule doubleLiteral, Num::Float
+        rule decimalLiteral, Num::Float
+        rule digits, Num
+        rule stringLiteral, Literal::String
+
+        # Variables
+        rule /\$/, Name::Variable, :varname
+
+        # Operators
+        rule operators, Operator
+        rule /\b#{word_operators}\b/, Operator::Word
+        rule /\b#{keywords}\b/, Keyword
+        rule /[?,{}()\[\]]/, Punctuation
+
+        # Functions
+        rule /(function)(\s*)(#{openParen})/ do # function declaration
+          groups Keyword, Text, Punctuation
+        end
+        rule /(map|array|empty-sequence)/, Keyword # constructors
+        rule /(#{kindTest})(\s*)(#{openParen})/ do  # kindtest
+          push :kindtest
+          groups Keyword, Text, Punctuation
+        end
+        rule /(#{kindTestForPi})(\s*)(#{openParen})/ do # processing instruction kindtest
+          push :kindtestforpi
+          groups Keyword, Text, Punctuation
+        end
+        rule /(#{eqName})(\s*)(#{openParen})/ do # function call
+          groups Name::Function, Text, Punctuation
+        end
+        rule /(#{eqName})(\s*)(#)(\s*)(#{digits})/ do # namedFunctionRef
+          groups Name::Function, Text, Name::Function, Text, Name::Function
+        end
+
+        # Type commands
+        rule /(cast|castable)(\s+)(as)/ do
+          goto :singletype
+          groups Keyword, Text, Keyword
+        end
+        rule /(treat)(\s+)(as)/ do
+          goto :itemtype
+          groups Keyword, Text, Keyword
+        end
+        rule /(instance)(\s+)(of)/ do
+          goto :itemtype
+          groups Keyword, Text, Keyword
+        end
+        rule /\b(as)\b/ do
+          token Keyword
+          goto :itemtype
+        end
+
+        # Paths
+        rule /\.\.|\.|\*/, Operator
+        rule /(#{ncName})(\s*)(:)(\s*)(\*)/ do
+          groups Name::Tag, Text, Punctuation, Text, Keyword::Reserved
+        end
+        rule /(\*)(\s*)(:)(\s*)(#{ncName})/ do
+          groups Keyword::Reserved, Text, Punctuation, Text, Name::Tag
+        end
+        rule /(#{axes})(\s*)(::)/ do
+          groups Keyword, Text, Operator
+        end
+        rule /@/, Name::Attribute, :attrname
+        rule eqName, Name::Tag
+
+        # Whitespace
+        rule /(\s+)/m, Text
+      end
+
+      state :singletype do
+        # Whitespace and comments
+        rule /\s+/m, Text
+        rule commentStart, Comment, :comment
+
+        # Type name
+        rule eqName do
+          token Keyword::Type
+          goto :root
```

Comment:
It's more idiomatic in Rouge to pop states. [`goto`](https://github.com/rouge-ruby/rouge/blob/88a40094cc2bac506dcab20aad4bd4ec45491b7f/lib/rouge/regex_lexer.rb#L407-L413) is used for situations where you want to move from one state to another by replacing the former state with the latter state. Doing that for `:root` will result in multiple `:root` states being on the stack.

Reply:
I think this is actually fine. The only states that are pushed onto the stack (rather than gone to with `goto`) are:

- `:comment`
- `:varname`
- `:attrname`
- `:kindtest`
- `kindtestforpi`

None of these states use `goto`. They all either pop, or push another state from this list. The only way to reach a `goto` statement is thus to be at the bottom of the stack (in other words, a stack of height one), so we shouldn't be able to have multiple `:root` states.

The reason I used `goto` so extensively is because it was easier to stay close to the [lexing instructions in this W3C document](https://www.w3.org/TR/xquery-xpath-parsing/), which do not always use the stack.

Reply:
My explanation here wasn't very good and the terminology of Rouge is a little confusing.

There is no way to avoid the stack in Rouge. When you use `goto`, what you're doing is not avoiding the stack but swapping out the current state on the stack for the target state (in all honesty the `goto` name should probably be deprecated and we refer to instead as `swap` so that this would be clearer). You can see that in the method definition:

https://github.com/rouge-ruby/rouge/blob/27d08a3af48153c5afdb06a0c38c1c68834e07db/lib/rouge/regex_lexer.rb#L408-L413

So while you're correct that you don't actually push many states onto the stack with this lexer, this is in itself a problem. The root state should always be, well, the root state. Removing it from the stack could lead to problems.

Reply:
I see your point, and made some changes in a90fe4452bc0cdc1d15fbf6bfd88ed958c38b05e. I've ensured that `:root` stays at the bottom of the stack at all times

For instance, before this commit:

```
stack: [:root]
stack: [:itemtype]
stack: [:itemtype]
stack: [:occurrenceindicator]
```


After:

```
stack: [:root]
stack: [:root, :itemtype]
stack: [:root, :itemtype]
stack: [:root, :occurrenceindicator]
```

I haven't replaced all `goto`s, because the specification specifies some state transitions as regular FSM transitions, without the need for a stack. I'm afraid that diverging from the W3C spec too much will lead to subtle bugs. If you're okay with this more restricted usage of goto, combined with regular stack usage, then I think this currently is a good solution :)